### ISSUE TITLE
Weird behavior with getProductsForSale()

### ISSUE THREAD

**Issue by harrisrobin** [iaphub/react-native-iaphub#44]
Hi all,

I'm running into an issue where when I call:
```js
Iaphub.getProductsForSale()
```
I correctly get a network call that has the products, however the function returns an empty array. I have published a listing, not sure why im getting an empty array when i do see the product in the network call that this library is performing.

Please see screenshot below.
 
![image](https://user-images.githubusercontent.com/3880008/102391890-30671c80-3f8b-11eb-8759-24dc7fb59a7b.png)

### P.S: Interestingly, this library performs two calls (not sure why). Here's the second which responds with nothing:

![image](https://user-images.githubusercontent.com/3880008/102392071-72905e00-3f8b-11eb-8f17-6e78bfea098b.png)

Any help is appreciated 

## TLDR: 
1. Calling `Iaphub.getProductsForSale()`
2. Getting an empty array
3. Seeing that the products are being fetched in network call but still getting an empty array when using getProductsForSale

**Comment by iaphub** [comment]
Hi @harrisrobin,

Are you seeing an error "Product sku not found" in the console?
When the product isn't returned by StoreKit it is filtered from the products returned by the getProductsForSale method.
Your sandbox environment is probably not configured properly (https://iaphub.com/docs/set-up-ios/configure-sandbox-testing)

Regarding the second API call it is normal, it is the API call reporting the pricing of the products.

**Comment by scaralfred** [comment]
Having the same issue. Did you solve it?

**Comment by scaralfred** [comment]
I solved it... it was a matter of publishing the products in V2 rather than V1.

**Comment by dancixx** [comment]
On iOS need to create Storekit environment? If i have storekit file i get the product from it and the sandbox purchase is successful but got an error after it, but when i delete storekit got nothing products.

**Comment by iaphub** [comment]
@dancixx, you don't have to use a storekit file, the storekit file gives the ability to emulate products on the client without loading the products from ItunesConnect.
But when the storekit file is used it isn't possible to validate the receipt with Itunes so IAPHUB won't be able to process your purchase.
If you do not get your products without your storekit file it means your itunes account or sandbox environment isn't configured properly (so itunes isn't returning your products).
Double check you did all the steps listed here: https://iaphub.com/docs/set-up-ios/configure-sandbox-testing
It is common for people to forget to sign all the itunes agreements, also a free Apple Developer Account won't work you need to enroll to the program.

**Comment by dancixx** [comment]
Thx for help, the problem solved. One more question i dont want to open new issue, so is it normal that the purchase method is much slower on ios then android?

**Comment by iaphub** [comment]
Great!
Yes it is a little bit slower on ios, that's itunes there is nothing we can do about it..
It is also slower in sandbox compared to the production in my opinion.

**Comment by Marcoo09** [comment]
Hi @iaphub @dancixx! 
I'm having the same problem. 
Now I'm testing on a real device with IOS 12.5.1 and I'm not able to setup the sandbox account.
I tried to getProductsForSale and I receive an empty array.
Then I tried to buy a product which I know the sku and I receive the following error:
![image](https://user-images.githubusercontent.com/18408823/112997673-b3c80a80-9143-11eb-82b6-e240e38237cd.png)
In any of the scenarios, I didn't find the section of sandbox account on settings of the device.
PS: I followed the guide that you share to configure sandbox account on ios and I'm not logged with any apple id on the device.

Could you please tell me how can I set the sandbox account to test the complete flow of a purchase? We're stuck on that.

Thanks in advance!

**Comment by dancixx** [comment]
Hi i have been written my exp in this question in this section:
https://github.com/dooboolab/react-native-iap/issues/1200

Maybe try this first then will not work i help you to solve the issue. :)

**Comment by nicoache1** [comment]
I @iaphub @dancixx I have a similar problem my issue is that i cannot find the Apple Sandbox Tester section in App Store (My device uses iOS 14) so i cannot retreive the products from IAPHub and cannot made any purchases. There is some kind of workaround or something?

**Comment by dancixx** [comment]
You have to in settings/appstore and the last item is the sandbox user. First maybe you should set up your phone as developer.

**Comment by nicoache1** [comment]
Yeah, i go to that but that last item doesnt appear at all. Also my device has inside settings the developer option so i think its already setted up.

**Comment by Marcoo09** [comment]
@dancixx I see the other issue but I continue stucked. I also don't understand clear what mean when was suggested 'Resending the application with updated prices with details and identified the Apple sandbox user.'
Where is done the process of identify the apple sandbox user?
If I test on testflight is not necessary to do that?
Thank you for your fast response 💪

**Comment by dancixx** [comment]
You have to create a Sandbox user in this section of appstoreconnect.

![Uploading Képernyőfotó 2021-03-30 - 23.52.40.png…]()

**Comment by iaphub** [comment]
@Marcoo09, your issue is definitely due to the sandbox environment not being configured correctly.
Did you configure your app on IAPHUB correctly ? (Products and listing created?)
Also please double check this page to configure your sandbox environment: https://www.iaphub.com/docs/set-up-ios/configure-sandbox-testing.
A common issue is to have the iTunes Agreements not accepted, when they are not accepted the sandbox environment won't work.

@nicoache1, when your sandbox environment is configured correctly you should be able to load the products for sale (even if you're not connected to the sandbox account).
If it is not configured correctly and the product isn't returned by iTunes you should see an error in the console.
After you're able to load the products, purchase a product and enter your sandbox account credentials and the sandbox section in the settings should appear after the 1st purchase.

**Comment by Marcoo09** [comment]
Thanks @dancixx @iaphub !

**Comment by krose422** [comment]
Hi @iaphub I am having the hardest time getting past `Error: Network Error` when I request products for sale. I followed every step to set up the Sandbox environment, and had actually been getting the array of products two days ago. But since then, I haven't been able to get anything despite not changing any configuration. I am setting things up for both iOS and Android. 

I can confirm the following: 

- [ x ] Agreements signed & banking / tax set up
- [ x ] Sandbox user is created and added to the Testers under Sandbox section in Users & Access
- [ x ] I am signed in with the sandbox apple id on my device (no other apple id is being used on the device)
- [ x ] I have created a group and products within that group in App Store connect and each are cleared for sale and have a localization
- [ x ] The bundle identifier set up in IAPHUB is correct
- [ x ] The In App Purchases capability is enabled in the configuration of the project

Any additional troubleshooting steps or help would be greatly appreciated. Thank you!

**Comment by krose422** [comment]
This actually started working for me again!
